### PR TITLE
feat: Adding support for CustomTabs and fix ASWebAuthenticationSession in WebAuthenticationBroker

### LINKED ISSUE #9338: Update WebAuthenticationBroker to make use of CustomTabs (Android) and ASWebAuthenticationSession (iOS)
## What would you like to be added:

Android implementation of WAB should use CustomTabs
iOS implementation should use ASWebAuthenticationSession  and fall back to SFAuthenticationSession

## Why is this needed:

WAB should make use of latest guidance for Android and iOS

## For which Platform:

- [ X ] iOS
- [ X ] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

### PR DESCRIPTION
**Pull request by nickrandolph**
GitHub Issue (If applicable): closes #9338

## PR Type

What kind of change does this PR introduce?
- Bugfix
- Feature

## What is the current behavior?

Android: Doesn't use customtabs which is now the recommended model for doing web based authentication in an app
iOS: ASWebAuthenticationSession throws exception in contructor when a full url is used as callbackuri (ie scheme://path instead of just scheme )

## What is the new behavior?

Android: Now uses fall back from CustomTabs (if available) to external browser
iOS: ASWebAuthenticationSession uses callbackuri.scheme instead of the whole callbackuri

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ X ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ X ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ X ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/SamplesApp/SamplesApp.UWP.Design/SamplesApp.UWP.Design.csproj [hunk 1]
```diff
@@ -131,9 +131,7 @@
     <PackageReference Include="Microsoft.NETCore.UniversalWindowsPlatform">
       <Version>6.2.8</Version>
     </PackageReference>
-    <PackageReference Include="Newtonsoft.Json">
-      <Version>12.0.3</Version>
-    </PackageReference>
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.UWP.Design/SamplesApp.UWP.Design.csproj:134]: Let's change this globally in the Directory.Build.targets

File: src/SamplesApp/UITests.Shared/Windows_Security/Authentication_Web/AuthenticationBroker_Demo.xaml.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ private async void PrepareClient()
 				// Create options for endpoint discovery
 				var options = new OidcClientOptions()
 				{
-					Authority = "https://demo.identityserver.io",
+					Authority = "https://demo.duendesoftware.com/",
```
Comment (jeromelaban) [inline src/SamplesApp/UITests.Shared/Windows_Security/Authentication_Web/AuthenticationBroker_Demo.xaml.cs:36]: Why this one ? (how did you choose it and what should be done next time this new one does not work)

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBroker.Android.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+﻿#nullable enable
+using Android.Content;
+using Uno.AuthenticationBroker;
+
+namespace Windows.Security.Authentication.Web
+{
+	public static partial class WebAuthenticationBroker
+	{
+		public static bool OnResume(Intent? intent=null)
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBroker.Android.cs:9]: Does it need to be public ?

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerActivityBase.Android.cs [hunk 4]
```diff
@@ -14,9 +17,81 @@ protected override void OnCreate(Bundle? savedInstanceState)
...
+
+				// finish this activity
+				Finish();
+			}
+		}
+	}
+
+
+	[Activity(ConfigurationChanges = ConfigChanges.Orientation | ConfigChanges.ScreenSize, Exported = true)]
+	internal class WebAuthenticationBrokerRedirectActivity : Activity
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerActivityBase.Android.cs:34]: Let's place this in a separate file

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerActivityBase.Android.cs [hunk 5]
```diff
@@ -14,9 +17,81 @@ protected override void OnCreate(Bundle? savedInstanceState)
...
+		Intent? actualIntent;
+
+		protected override void OnCreate(Bundle? savedInstanceState)
+		{
+			base.OnCreate(savedInstanceState);
+
+			var extras = savedInstanceState ?? Intent?.Extras;
+
+			// read the values
+			launched = extras?.GetBoolean(launchedExtra, false)??false;
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerActivityBase.Android.cs:49]:
```suggestion
			launched = extras?.GetBoolean(launchedExtra, false) ?? false;
```

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs [hunk 6]
```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
...
+				var intentUri = new Uri(intentData);
 
-		internal static void OnMainActivityResumed()
-		{
-			// This is called when the application activity is resumed.
-			// If it occurs before the TCS is competed, it means the user canceled the operation.
+				// Only handle schemes we expect
+				if (!CanHandleCallback(currentRedirectUri, intentUri))
+				{
+					tcsResponse.TrySetException(new InvalidOperationException($"Invalid Redirect URI, detected `{intentUri}` but expected a URI in the format of `{currentRedirectUri}`"));
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:94]: What should the user do if the exception happens? Can you add a "Verify this, do that, check this" in the exception text?

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs [hunk 7]
```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
...
 
-			await LaunchBrowserCore(options, requestUri, callbackUri, ct);
+			// Try to find the activity for the callback intent
+			if (!IsIntentSupported(intent, packageName))
+				throw new InvalidOperationException($"You must subclass the `{nameof(WebAuthenticationBrokerActivityBase)}` and create an IntentFilter for it which matches your `{nameof(callbackUrl)}`.");
 
-			var result = await tcs!.Task;
+			// Cancel any previous task that's still pending
+			if (tcsResponse?.Task != null && !tcsResponse.Task.IsCompleted)
+				tcsResponse.TrySetCanceled();
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:131]:
```suggestion
			if (tcsResponse?.Task != null && !tcsResponse.Task.IsCompleted)
			{
				tcsResponse.TrySetCanceled();
			}
```

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs [hunk 8]
```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
...
 
-			return result;
+			return await tcsResponse.Task;
 		}
 
-		protected virtual async Task LaunchBrowserCore(WebAuthenticationOptions options,
-			Uri requestUri,
-			Uri callbackUri,
-			CancellationToken ct)
+		private async Task<bool> StartCustomTabsActivity(Uri url)
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:147]: should you forward the `CancellationToken` here? (same in `BindServiceAsync`)

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs [hunk 9]
```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
...
+			}
+			finally
+			{
+				try
+				{
+					customTabsActivityManager.Client?.Dispose();
+				}
+				finally
+				{
+				}
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:180]: Do you need this finally?

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs [hunk 10]
```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
...
+			{
+				manager.CustomTabsServiceConnected -= OnCustomTabsServiceConnected;
+				tcs.TrySetResult(true);
+			}
+		}
+
+		internal static bool IsIntentSupported(Intent intent, string? expectedPackageName)
+		{
+			if (Application.Context is not Context ctx || ctx.PackageManager is not PackageManager pm)
+				return false;
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:210]:
```suggestion
			if (Application.Context is not Context ctx || ctx.PackageManager is not PackageManager pm)
			{
				return false;
			}
```

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs [hunk 11]
```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
...
+
+			return intent.ResolveActivity(pm) is ComponentName c && c.PackageName == expectedPackageName;
+		}
+
+		internal static bool CanHandleCallback(Uri expectedUrl, Uri callbackUrl)
 		{
-			await Launcher.LaunchUriAsync(requestUri);
+			if (!callbackUrl.Scheme.Equals(expectedUrl.Scheme, StringComparison.OrdinalIgnoreCase))
+				return false;
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:218]:
```suggestion
			if (!callbackUrl.Scheme.Equals(expectedUrl.Scheme, StringComparison.OrdinalIgnoreCase))
			{
				return false;
			}
```

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs [hunk 12]
```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
...
+
+			if (!string.IsNullOrEmpty(expectedUrl.Host))
+			{
+				if (!callbackUrl.Host.Equals(expectedUrl.Host, StringComparison.OrdinalIgnoreCase))
```
Comment (jeromelaban) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:222]: Should it be an ignore case comparison?

File: src/Uno.UWP/Uno.csproj [hunk 13]
```diff
@@ -52,11 +52,13 @@
 		<PackageReference Include="Xamarin.AndroidX.Legacy.Support.v4" />
 		<PackageReference Include="Xamarin.AndroidX.AppCompat" />
 		<PackageReference Include="Xamarin.AndroidX.Fragment" />
+		<PackageReference Include="Xamarin.AndroidX.Browser"  Version="1.0.0" />
```
Comment (jeromelaban) [inline src/Uno.UWP/Uno.csproj:55]: Taking a dependency here [requires the nuspec](https://github.com/unoplatform/uno/blob/6240b58a4d94e43bba3567763866a15de62041aa/build/Uno.WinUI.nuspec#L17) to include it as well.

File: src/Uno.UWP/Uno.net6.csproj [hunk 14]
```diff
@@ -53,6 +53,7 @@
 		<PackageReference Include="Xamarin.AndroidX.Legacy.Support.v4" />
 		<PackageReference Include="Xamarin.AndroidX.AppCompat" />
 		<PackageReference Include="Xamarin.AndroidX.Fragment" />
+		<PackageReference Include="Xamarin.AndroidX.Browser"  Version="1.0.0" />
```
Comment (jeromelaban) [inline src/Uno.UWP/Uno.net6.csproj:56]: Taking a dependency here [requires the nuspec](https://github.com/unoplatform/uno/blob/6240b58a4d94e43bba3567763866a15de62041aa/build/Uno.WinUI.nuspec#L17) to include it as well.

File: src/SamplesApp/SamplesApp.UWP.Design/SamplesApp.UWP.Design.csproj (hunk 1 above)
Comment (nickrandolph) [reply src/SamplesApp/SamplesApp.UWP.Design/SamplesApp.UWP.Design.csproj]: Done

File: src/SamplesApp/UITests.Shared/Windows_Security/Authentication_Web/AuthenticationBroker_Demo.xaml.cs (hunk 2 above)
Comment (nickrandolph) [reply src/SamplesApp/UITests.Shared/Windows_Security/Authentication_Web/AuthenticationBroker_Demo.xaml.cs:36]: IdentityServer from what I recall is now a commercial offering which is offered by DuendeSoftware. This demo endpoint is a service they offer: https://docs.duendesoftware.com/identityserver/v6/overview/demo_server/
If this demo server is decommissioned we may need to look at standing up our own instance for dev/testing

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBroker.Android.cs (hunk 3 above)
Comment (nickrandolph) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBroker.Android.cs:9]: Changed to internal

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs (hunk 9 above)
Comment (nickrandolph) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:180]: Removed

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs (hunk 12 above)
Comment (nickrandolph) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs:222]: https://www.rfc-editor.org/rfc/rfc3986 - indicates that scheme and host are case in-sensitive where as the rest of the uri is case sensitive.

File: src/Uno.UWP/Uno.net6.csproj (hunk 14 above)
Comment (nickrandolph) [reply src/Uno.UWP/Uno.net6.csproj:56]: done

File: src/Uno.UWP/Uno.csproj (hunk 13 above)
Comment (nickrandolph) [reply src/Uno.UWP/Uno.csproj:55]: done

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs (hunk 8 above)
Comment (nickrandolph) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs]: yes, done

File: src/SamplesApp/Directory.Build.props [hunk 15]
```diff
@@ -7,7 +7,11 @@
     <!-- fixed for self-testing inside the Uno Solution -->
     <UnoRemoteControlPort>53487</UnoRemoteControlPort>
   </PropertyGroup>
-
+  
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
+  </ItemGroup>
```
Comment (jeromelaban) [inline src/SamplesApp/Directory.Build.props:13]: We're generally updating the versions in `src/Directory.Build.props`, why did you need it scoped to the samples app? Is the updated version only used by the sample app, or exposed through Uno ?

**Review by jeromelaban — APPROVED** [review]

File: src/SamplesApp/Directory.Build.props (hunk 15 above)
Comment (nickrandolph) [reply src/SamplesApp/Directory.Build.props:13]: I only saw that the samples app was referencing Newtonsoft, so didn't push to the parent Directory.Build.props.
